### ISSUE TITLE
[ENH] Benchmarking Discussion

### ISSUE THREAD

**Issue by TNTran92** [alan-turing-institute/sktime#2777]
During the benchmarking meeting today, a number of topics was discussed regarding benchmarking. This issue summarizes/expands these discussions. This documents will be updated after meetings:
Three packages were among the discussions on 06/10/2022:
* Benchmarking-wise, the `Evaluator()` class of `gluonts` requires about 3 adapters to make it work with `sktime`. I do recall exactly but one of the adapter would be for sktime's `metrics`. Need to consider additional options. There are 2 candidates: Kotsu and OpenML
* Kotsu is a high level and light-weight package that performs benchmarking that has no dependency on specific packages. At a high level, it allows users to register a dataset with a predicting task and then apply estimators to that registration. Will need more investigation about application to sktime.
* OpenML performs a similar task to kotsu. An additional plus is that such registration has a share-able ID that can be shared on the cloud, allowing other users to repeat the same task. At the moment, it has limited support for time series dataset. Will also need to investigate its application in `sktime`.

**To do tasks (High priority):**
- [x] Investigate the implementation of `kotsu` into `sktime` - refer to Issue #2804
- [x] Investigate implementation of `OpenML` into sktime - refer to Issue #2821
- [x] Investigate implementaition of  `Mlflow` into sktime - refer to Issue #2916 
- [x] Compare available benchmarking approaches
- [ ] Converter for Multiindex DF into ListDataset -- refer to Issue #2860 
- [ ] Converter for ListDataset into Multiindex DF

Update: 6/20/2022
- [x] kotsu's experiment: Test all estimators for point forecasting
- [x]  kotsu's experiment: Test all estimators for interval forecasting
- [ ] kotsu's experiment: Investigate Issue #2742
- [ ] Prep M4 dataset for sktime
- [ ] Mock a benchmarking study with vanilla kotsu
- [ ] kotsu's experiment: Draft interface design for kotsu
- [ ] kotsu's experiment: Try with annotations

**Good to have (Lower priority):**
N/A

**Additional thoughts (Not action item)**
If implemented,  the `gluonts`'s adapter would look similar to statsmodel's adapter.

**Comment by ltsaprounis** [comment]
@DBCerigo - FYI for kotsu

**Comment by fkiraly** [comment]
@DBCerigo (also FYI @TNTran92, @ltsaprounis), we're basically tring to set up benchmarking functionality for things similar to "run the M4 or M5 competition with a specific set of estimators", or "perform rolling window backtesting with these estimators", or "replicate the time series classification bake-off with these estimators.

There is some historic work in `sktime`:

* in 2019/2020, @mloning recreated the M4/M5 competition for the `sktime` paper with a few methods. I don't think he ever shared the code, though.
* in 2019, @TonyBagnall also wanted to replicate his benchmarking studies from earlier papers, most prominently, the "time series classification bake-off". I think he never did that either, but there must be some code that his research group is using. I find it unlikely that we'll get access or that that moves into `sktime`. But being able to do sth similar would be a nice goal.
* the benchmarking module of `sktime`, which I think still works but has run out of sync a bit, design-wise https://github.com/alan-turing-institute/sktime/tree/main/sktime/benchmarking. @ViktorKaz wanted to work on this but that also got abandoned.

In summary, it seems that benchmarking is a topic which seems at the same time hard design-wise, and a topic where people are either reluctant or unable to share or produce clean, reusable code.

The goals for me, though, would be clear:

* few-liner to run M4/M5 like benchmarks for forecasting
* few-liner to run bake-off like benchmarks for classification

**Comment by TNTran92** [comment]
TLDR:
In the order of most  to least work
OpenML -> gluon-ts -> kotsu

kotsu works straight out of the box - we do not have to worry about adapters and just on how to incorporate it into sktime
Both OpenML and gluon-ts require adapters for dataset, estimator and `metrics.` Gluonts may not help us as a benchmarking framework but it has DL models and other tool that willl be nice-to-have.

**Comment by DBCerigo** [comment]
@fkiraly 
> @DBCerigo (also FYI @TNTran92, @ltsaprounis), we're basically tring to set up benchmarking functionality for things similar to "run the M4 or M5 competition with a specific set of estimators", or "perform rolling window backtesting with these estimators", or "replicate the time series classification bake-off with these estimators.

`kotsu` was designed exactly to do these things, with the following properties of minimizing bloat or overhead, allowing for any model interface/validation, and avoiding common pitfalls when aiming for fair benchmarks  [ref](https://github.com/datavaluepeople/kotsu#what-is-it). 

"run the M4 or M5 competition", and "perform rolling window backtesting" would be a `validation`s and each estimator would be a `model`. Register them and "hit" `run`. 

> In summary, it seems that benchmarking is a topic which seems at the same time hard design-wise, and a topic where people are either reluctant or unable to share or produce clean, reusable code. Maybe this is where `kotsu` (or openML etc) could help?

Again, we built `kotsu` specifically to solve those problems of benchmarking often becoming a messy, bloated, and even inaccurate (unfair comparisons) endeavours. 

> (and it should be flexible to change the data sets used, the metrics used, the estimators used)

Yep it is designed for such flexibility. 

@TNTran92 
> kotsu works straight out of the box - we do not have to worry about adapters and just on how to incorporate it into sktime

Glad to hear that.

From https://github.com/TNTran92/sktime/blob/kotsu_application/sktime/benchmarking/kotsu_sktime_application.ipynb it looks like you already know what you're doing.


Myself and co (@ali-tny @alex-hh) would be interested in contributing to/supporting on implementing such benchmarking. We use it for all our projects and so can offer our usual patterns/structure for using it in a clean and unobtrusive way. 

Are there any time constraints or aimed-for deadlines for this?

**Comment by fkiraly** [comment]
> Are there any time constraints or aimed-for deadlines for this?

We have a summer programme, which just started, and a workstream lifetime until Sep or Oct.

We will have developer days in London, July 11-15 - you seem to be in London, too, would you like to join?
There are also community/research talk slots, you could have one if you like?

(could also do this on one of the earlier days since it likely influences the work)

https://github.com/sktime/community-org/blob/main/events/sktime-dev-days-2022.md

**Comment by TNTran92** [comment]
> * add `kotsu` into `sktime` (unobtrusively)
> * adds a generalised validation factory func (probably takes a dataset, a cv-splitter, and a list of metrics)
> * register a few validations that we think are relevant (based on your messages above + @TNTran92's spike PR)
> * register a few estimators
> * module/docs/example on how to run the estimators through the validations, and read the stored the results

@DBCerigo  That is very kind of you. I am trying to get sktime's benchmarking done over the summer, so greatly appreciate inputs from the authors themselves. Do you mind if I sit in when you and co discuss sktime integration? We can invite other people from sktime side too.

> * I run a session on benchmarking using kotsu at the dev days on July 12th at 2pm (I'm travelling back to UK on 11th so this slot is earliest I can do). I can be contributing code/being around during more of that week also.

I'm from the US 🙃, so I will join the dev days online only.

**Comment by ltsaprounis** [comment]
Good stuff! Pinged you all in the sktime slack to organise a call.

**Comment by AurumnPegasus** [comment]
Hey @TNTran92 ,
I was wondering if I could help by taking on the fourth task

> Need to implement gluonts's estimators/functionalities later on. Make a converter that convert ListDataset into Hierarchical type in sktime.
>

Maybe I can start of by making the converter if its not done?

**Comment by TNTran92** [comment]
> Hey @TNTran92 , I was wondering if I could help by taking on the fourth task
> 
> > Need to implement gluonts's estimators/functionalities later on. Make a converter that convert ListDataset into Hierarchical type in sktime.
> 
> Maybe I can start of by making the converter if its not done?

PR #2877 tackles the first part of this adapter. You can go ahead and use it while the merge is going through

**Comment by mloning** [comment]
> * in 2019/2020, @mloning recreated the M4/M5 competition for the `sktime` paper with a few methods. I don't think he ever shared the code, though. I don't think I have access to it, for instance.

FYI the code repo is referenced in the [paper](https://arxiv.org/pdf/2005.08067.pdf) and publicly available here: https://github.com/mloning/sktime-m4

**Comment by fkiraly** [comment]
>FYI the code repo is referenced in the [paper](https://arxiv.org/pdf/2005.08067.pdf) and publicly available here: https://github.com/mloning/sktime-m4

Ah, there it is! Nice!!
I already feared it was lost.

FYI @DBCerigo
